test-data/unit/check-unions.test

```diff
@@ -1048,3 +1048,17 @@ def foo(a: T2, b: T2) -> T2:
 def bar(a: T4, b: T4) -> T4:  # test multi-level alias
     return a + b
 [builtins fixtures/ops.pyi]
+
+
+[case testNarrowingUnionFromInContainer]
+from typing import Union, NewType, Set
+MyCoolInt = NewType('MyCoolInt', int)
+T1 = Union[int, float, MyCoolInt, str]
+
+my_cool_set: Set[Union[int, str]] = set(2, 'hi')
+
+def my_cool_function(my_cool_arg: T1):
+    if my_cool_arg in my_cool_set:
+        reveal_type(my_cool_arg)  # N: Revealed type is Union[int, str]
```

Comment:
This is not quite safe; for example, `2.0 in {2}` returns True, so it could still be a float.

Reply:
Yeah and now that I think about it, it wouldn't necessarily be safe if the class's `__contains__` or `__eq__` was doing something special. Is that something mypy tries to help with?

Reply:
Oh I think I see, this should borrow some of the same precondition logic that the `==`/`is` checks use like ten lines above it